### PR TITLE
QemuQ35Pkg: Integrate TPM Replay feature

### LINKED ISSUE #677: [Feature]: Integrate the TPM Replay feature
### Feature Overview

A new feature was recently added to [mu_plus](https://github.com/Microsoft/mu_plus) called TPM Replay.

Details are in https://github.com/microsoft/mu_plus/pull/287.

That feature is very useful on a virtual platform like QEMU because it allows rapid testing of TPM Replay event logs replayed by the feature. This issue tracks integrating the feature in `QemuQ35Pkg`.

### Solution Overview

- Add the necessary libraries and drivers to the DSC and FDF behind a `TPM_REPLAY_ENABLED` flag.
- Provide a PEI instance of the `FvMeasurementExclusionLib` library class for the platform.
- Add QEMU-specific documentation describing the how the feature works and is used in QEMU.

### Alternatives Considered

_No response_

### Urgency

Medium

### Are you going to implement the feature request?

I will implement the feature

### Do you need maintainer feedback?

No maintainer feedback needed

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by makubacki**
Closes #677 

## Description

This feature enables experimentation with controlling the TPM
measurements made by firmware. It allows a developer to craft
a TPM event log and have the replayed during boot. It is disabled
by default.

General feature information is available in its readme:

https://github.com/microsoft/mu_plus/blob/HEAD/TpmTestingPkg/TpmReplayPei/Readme.md

It is currently only planned to be used QemuQ35Pkg. It could be
enabled in QemuSbsaPkg in the future if there is interest.

Details about the feature in the context of QEMU, are described in
`Platforms/Docs/Q35/Features/feature_tpm_replay.md` in this change.

- [x] Impacts functionality?
  - **Functionality** - Does the change ultimately impact how firmware functions?
  - Examples: Add a new library, publish a new PPI, update an algorithm, ...
- [ ] Impacts security?
  - **Security** - Does the change have a direct security impact on an application,
    flow, or firmware?
  - Examples: Crypto algorithm change, buffer overflow fix, parameter
    validation improvement, ...
- [ ] Breaking change?
  - **Breaking change** - Will anyone consuming this change experience a break
    in build or boot behavior?
  - Examples: Add a new library class, move a module to a different repo, call
    a function in a new library class in a pre-existing module, ...
- [ ] Includes tests?
  - **Tests** - Does the change include any explicit test code?
  - Examples: Unit tests, integration tests, robot tests, ...
- [x] Includes documentation?
  - **Documentation** - Does the change contain explicit documentation additions
    outside direct code modifications (and comments)?
  - Examples: Update readme file, add feature readme file, link to documentation
    on an a separate Web page, ...

## How This Was Tested

- Input YAML to binary
- Input JSON to binary
- Input binary to YAML
- Replay events on QEMU Q35 to PCRs
- Verify event log in OS against the input file

## Integration Instructions

See `feature_tpm_replay.md`.

Enable the feature in `QemuQ35Pkg` in the DSC file:

```ini
  DEFINE TPM_REPLAY_ENABLED = TRUE
```

Or, as a `stuart_build` argument:

```bash
> stuart_build -c Platforms/QemuQ35Pkg/PlatformBuild.py --flashrom TOOL_CHAIN_TAG=GCC5 BLD_*_TPM_ENABLE=TRUE \
               BLD_*_TPM_ENABLE=TRUE TPM_DEV=/tmp/mytpm1/swtpm-sock
```

Then, follow instructions to create the TPM Replay
event log and load it.

### CODE REVIEW

File: Platforms/Docs/Q35/Features/feature_tpm_replay.md [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+```ini
+  DEFINE TPM_REPLAY_ENABLED = TRUE
+```
+
+Or, as a `stuart_build` argument:
+
+```bash
+> stuart_build -c Platforms/QemuQ35Pkg/PlatformBuild.py --flashrom TOOL_CHAIN_TAG=GCC5 BLD_*_TPM_ENABLE=TRUE \
+               BLD_*_TPM_ENABLE=TRUE TPM_DEV=/tmp/mytpm1/swtpm-sock
```
Comment (os-d) [inline Platforms/Docs/Q35/Features/feature_tpm_replay.md:25]:
```suggestion
               TPM_DEV=/tmp/mytpm1/swtpm-sock
```

You have `BLD_*_TPM_ENABLE=TRUE` twice.

File: Platforms/QemuQ35Pkg/QemuQ35Pkg.dsc [hunk 2]
```diff
@@ -924,6 +925,8 @@
   !include CryptoPkg/Driver/Bin/CryptoDriver.inc.dsc
 !endif
 
+QemuQ35Pkg/Library/PeiFvMeasurementExclusionLib/PeiFvMeasurementExclusionLib.inf
```
Comment (os-d) [inline Platforms/QemuQ35Pkg/QemuQ35Pkg.dsc:928]: Should this be gated by TPM_REPLAY_ENABLED? Saves some build time (not much), but also may trigger reporting tools that check for built binaries being used?

**Review by os-d — APPROVED** [review]

**Review by apop5 — APPROVED** [review]

File: Platforms/QemuQ35Pkg/QemuQ35Pkg.dsc (hunk 2 above)
Comment (makubacki) [reply Platforms/QemuQ35Pkg/QemuQ35Pkg.dsc:928]: I decided to keep this one in the normal platform build. It has a bit more platform touch points (like FV PCDs) than code being built in TpmTesting so I didn't want it to get broken and not discovered until the next person turns on the feature.

File: Platforms/Docs/Q35/Features/feature_tpm_replay.md (hunk 1 above)
Comment (makubacki) [reply Platforms/Docs/Q35/Features/feature_tpm_replay.md:25]: Thanks. Fixed in the latest push.

**Pull request merged**